staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go

```diff
@@ -429,58 +535,252 @@ func TestMuxAndDiscoveryComplete(t *testing.T) {
 }
 
 func TestPreShutdownHooks(t *testing.T) {
-	s := newGenericAPIServer(t, true)
-	doer := setupDoer(t, s.SecureServingInfo)
+	tests := []struct {
+		name   string
+		server func() *GenericAPIServer
+	}{
+		{
+			name: "ShutdownSendRetryAfter is disabled",
+			server: func() *GenericAPIServer {
+				return newGenericAPIServer(t, &fakeAudit{}, false)
+			},
+		},
+		{
+			name: "ShutdownSendRetryAfter is enabled",
+			server: func() *GenericAPIServer {
+				return newGenericAPIServer(t, &fakeAudit{}, true)
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			s := test.server()
+			doer := setupDoer(t, s.SecureServingInfo)
+
+			preShutdownHookErrCh := make(chan error)
+			err := s.AddPreShutdownHook("test-backend", func() error {
+				// this pre-shutdown hook waits for the shutdown duration to elapse,
+				// and then send a series of requests to the apiserver, and
+				// we expect these series of requests to be completed successfully
+				<-s.lifecycleSignals.AfterShutdownDelayDuration.Signaled()
+
+				// we send 5 requests, once every second
+				var err error
+				client := newClient(true)
+				for i := 0; i < 5; i++ {
+					r := doer.Do(client, func(httptrace.GotConnInfo) {}, fmt.Sprintf("/echo?message=attempt-%d", i), 1*time.Second)
+					err = r.err
+					if err == nil && r.response.StatusCode != http.StatusOK {
+						err = fmt.Errorf("did not get status code 200 - %#v", r.response)
+						break
+					}
+					time.Sleep(time.Second)
+				}
+				preShutdownHookErrCh <- err
+				return nil
+			})
+			if err != nil {
+				t.Fatalf("Failed to add pre-shutdown hook - %v", err)
+			}
 
-	preShutdownHookErrCh := make(chan error)
-	err := s.AddPreShutdownHook("test-backend", func() error {
-		// this pre-shutdown hook waits for the requests in flight to drain
-		// and then send a series of requests to the apiserver, and
-		// we expect these series of requests to be completed successfully
-		<-s.lifecycleSignals.InFlightRequestsDrained.Signaled()
-
-		// we send 5 requests, once every second
-		var r result
-		client := newClient(true)
-		for i := 0; i < 5; i++ {
-			r = doer.Do(client, func(httptrace.GotConnInfo) {}, fmt.Sprintf("/echo?message=attempt-%d", i), 1*time.Second)
-			if r.err != nil {
-				break
+			// start the API server
+			stopCh, runCompletedCh := make(chan struct{}), make(chan struct{})
+			go func() {
+				defer func() {
+					// this test has an inherent race condition when we wait for two go routines
+					// to finish - the Run method and the pre-shutdown hook, each running in
+					// its own goroutine, give it a second before unblocking the test assert
```

Comment:
This is asking for flakes in heavily starved environments.
If you want to ensure that actually this one will be called as a second one, let's synchronize it explicitlly.

Reply:
yeah, i don't remember why it was done this way. I changed the test, it should be flake free without a sleep now.